src/util/sighash.rs

```diff
@@ -87,6 +89,29 @@ pub struct ScriptPath<'s> {
     leaf_version: u8,
 }
 
+/// Hashtype of an input's signature, encoded in the last byte of the signature
+/// Fixed values so they can be casted as integer types for encoding
+#[derive(PartialEq, Eq, Debug, Copy, Clone)]
+pub enum SigHashType {
```

Comment:
I think this should be a 2 level enum where we have
```
enum SigHashType {
    Legacy,
    SegwitV0,
    SegwitV1,
}
```
to make clear contexts where modes do not have/have different meanings.

Reply:
I think we should just have different types of `SigHashTypeLegacy`, `SighashTypeSegwitv0`, `SigHashTypeTaproot. 

In what use case would you need a level 2 enum?

Reply:
I think that's OK -- I mean more that there should not be a "SigHashType" general enum that From coerces into other SigHashType* things. Espeically once we get into BIP-118, it would switch into a TryFrom because certain SigHashTypes would not be proper.

Reply:
> I think we should just have different types of SigHashTypeLegacy, SighashTypeSegwitv0, SigHashTypeTaproot.

why different types for legacy and segwit?

Is it a very bad idea to add a `RESERVED` variant to the `sighash::SigHashType` introduced in this PR to accommodate SIGHASH_PREVOUT when it's time?
`#[non_exhaustive]` would be better but it's not in our MSRV

Reply:
Sorry, I think I misread your comment initially for something else. I am not against adding the reserved variant, but I don't know if how effectively it serves the same purpose as `#[non-exhaustive]`. 

I will let other contributors comment on this, I am not sure if we should so much forward-thinking about `OP_ANYPREVOUT` right now. The current spec for BIP118 is itself a major change in PublicKey API because it introduces new key types. Maybe additionally doing this is that not that bad?

Reply:
I guess if I can better specify my position, I think that that one should be able to make a "pure" function:

```rust
fn get_sighash(tx:&TransactionData, sighash: SigHashType) -> Hash
```

which computes the correct digest with no additional instructions. Since legacy, segwitv0, and segwitv1 all compute different things, unifying the type as one `SigHashType` means that we need additional context, so

```rust
fn get_sighash_legacy(tx:&TransactionData, sighash: SigHashType) -> Hash
fn get_sighash_v0(tx:&TransactionData, sighash: SigHashType) -> Hash
fn get_sighash_v1(tx:&TransactionData, sighash: SigHashType) -> Hash
``` 

This is already sort of the case because v1 requires more context to sign, but I imagine the APIs will converge on having the superset of data available for any case (v1/v0/legacy).  I think you in theory could parse what digest version is required from the TxData, but it'd be nice if SigHashType were an explicit & unique digest instruction.

Reply:
I do agree this isn't a huge issue for now, just sounding the warning that BIP-118's plans do include diverging sighash types where there will be different things available at different times, and I think it is something we can handle relatively elegantly by being more explicit here.

Reply:
`non_exhaustive` does not work on current MSRV